repository/src/main/java/org/alfresco/repo/domain/schema/script/DeleteNotExistsV3Executor.java

```diff
@@ -0,0 +1,437 @@
+
+                    hasResults = primaryPrepStmt.execute();
+                }
+
+                // Check if we have any more ids to delete
+                if (!deleteIds.isEmpty())
+                {
+                    deleteFromPrimaryTable(deletePrepStmt, deleteIds, primaryTableName);
+                    connection.commit();
+                }
+            }
+        }
+        finally
+        {
+            closeQuietly(deletePrepStmt);
+            closeQuietly(secondaryPrepStmts);
+            closeQuietly(primaryPrepStmt);
+
+            closeQuietly(connection);
+        }
+
+        if (pauseAndRecover)
+        {
+            pauseAndRecoverJob(dataSource);
+            logger.info("Resuming the job on primary table " + primaryTableName + " picking up after id " + primaryId);
+            processBatch(primaryTableName, primaryColumnName, primaryWhereClause, primaryId, tableColumn, tableUpperLimits,
+                    optionalWhereClauses);
+        }
+    }
+
+    @Override
+    protected String createPreparedSelectStatement(String tableName, String columnName, String whereClause)
+    {
+        StringBuilder sqlBuilder = new StringBuilder("SELECT " + columnName + " FROM " + tableName + " WHERE ");
+
+        if (whereClause != null && !whereClause.isEmpty())
+        {
+            sqlBuilder.append(whereClause + " AND ");
+        }
+
+        sqlBuilder.append(
+                columnName + " >= ? AND " + columnName + " <= ? GROUP BY " + columnName + " ORDER BY " + columnName + " ASC ");
+
+        if (dialect instanceof MySQLInnoDBDialect)
+        {
+            sqlBuilder.append(" LIMIT " + batchSize);
+        }
+        else
+        {
+            sqlBuilder.append(" OFFSET 0 ROWS FETCH FIRST " + batchSize + " ROWS ONLY");
+        }
+
+        return sqlBuilder.toString();
+    }
+
+    @Override
+    protected Long processPrimaryTableResultSet(PreparedStatement primaryPrepStmt, PreparedStatement[] secondaryPrepStmts,
+            PreparedStatement deletePrepStmt, Set<Long> deleteIds, String primaryTableName, String primaryColumnName,
+            Pair<String, String>[] tableColumn) throws SQLException
+    {
+        int rowsProcessed = 0;
+        Long primaryId = null;
+        Long minSecId = 0L;
+        Long maxSecId = 0L;
+
+        // Get the potential IDS to delete from the primary table
+        Set<Long> potentialIdsToDelete = new HashSet<Long>();
+        try (ResultSet resultSet = primaryPrepStmt.getResultSet())
+        {
+            while (resultSet.next())
+            {
+                primaryId = resultSet.getLong(primaryColumnName);
+                potentialIdsToDelete.add(primaryId);
+            }
+        }
+
+        if (potentialIdsToDelete.size() == 0)
+        {
+            // Nothing more to do
+            return primaryId;
+        }
+
+        Long minPotentialId = Collections.min(potentialIdsToDelete);
+        Long maxPotentialId = Collections.max(potentialIdsToDelete);
```

Comment:
I don't know how large this set of potential ids is, but we could compute the min and max during creation of the set above.  Maybe this optimisation is completely negligible though